pandas/src/inference.pyx

```diff
@@ -655,22 +731,60 @@ def maybe_convert_numeric(object[:] values, set na_values,
         val = values[i]
 
         if val.__hash__ is not None and val in na_values:
+            seen_null = True
+            if _check_uint64_nan(seen_uint, seen_null,
```

Comment:
maybe this might be a bit simpler if instead of individual values we have a structure with those values instead, e.g.

```
struct Seen {
    bint null,
    bint uint, 
   .....
} seen;
```

then you can pass this to functions as needed (so the args to those checking functions will be consistent),
e.g. you don't have seen_uint, seen_null (and other args in another case).

can be done in a followup as well (if it makes sense)